pkg/grpc/api.go

```diff
@@ -368,6 +370,105 @@ func (a *api) GetSecret(ctx context.Context, in *runtimev1pb.GetSecretRequest) (
 	return response, nil
 }
 
+func (a *api) ExecuteStateTransaction(ctx context.Context, in *runtimev1pb.ExecuteStateTransactionRequest) (*empty.Empty, error) {
+	if a.stateStores == nil || len(a.stateStores) == 0 {
+		return &empty.Empty{}, errors.New("ERR_STATE_STORE_NOT_CONFIGURED")
+	}
+
+	storeName := in.StoreName
+
+	if a.stateStores[storeName] == nil {
+		return &empty.Empty{}, errors.New("ERR_STATE_STORE_NOT_FOUND")
+	}
+
+	transactionalStore, ok := a.stateStores[storeName].(state.TransactionalStore)
+	if !ok {
+		return &empty.Empty{}, errors.New(incompatibleStateStore)
+	}
+
+	requests := []state.TransactionalRequest{}
+	for _, inputReq := range in.Requests {
+		switch state.OperationType(inputReq.OperationType) {
+		case state.Upsert:
+			setReq := state.SetRequest{
+				Key:      a.getModifiedStateKey(inputReq.States.Key),
+				Metadata: inputReq.States.Metadata,
+				Value:    inputReq.States.Value,
+				Options:  a.getSetStateOptions(inputReq.States),
+			}
+			req := state.TransactionalRequest{
+				Operation: state.Upsert,
+				Request:   setReq,
+			}
+			requests = append(requests, req)
+
+		case state.Delete:
+			delReq := state.DeleteRequest{
+				Key:      a.getModifiedStateKey(inputReq.States.Key),
+				Metadata: inputReq.States.Metadata,
+				Options:  a.getDeleteStateOption(inputReq.States),
+			}
+			req := state.TransactionalRequest{
+				Operation: state.Delete,
+				Request:   delReq,
+			}
+			requests = append(requests, req)
+		default:
+			return &empty.Empty{}, fmt.Errorf("ERR_OPERATION_NOT_SUPPORTED: operation type %s not supported", inputReq.OperationType)
+		}
+	}
+
+	err := transactionalStore.Multi(requests)
+	if err != nil {
+		return &empty.Empty{}, fmt.Errorf("ERR_STATE_TRANSACTION: %s", err)
+	}
+	return &empty.Empty{}, nil
+}
+
+func (a *api) getSetStateOptions(r *commonv1pb.StateItem) (o state.SetStateOption) {
+	if r.Options != nil {
```

Comment:
this is the exactly duplicated implementation with getDeleteStateOption. Do you have better idea to have single implementation?

Can you please add unit-tests ?

Reply:
I added them in grpc/api_test.go

Reply:
Thanks!

Reply:
hmm I cannot find the test to test `getSetStateOptions`... maybe I miss it.

Reply:
Can't seem to test the getSetStateOptions because the api interface/struct in `api_test.go` does not have the method?

Reply:
you can refer this code - https://github.com/dapr/dapr/blob/e92c3d468d94f4954eb45ad8333084fa5a5f831e/pkg/grpc/api_test.go#L262-L265

you can create object like below and access `getSetStateOptions`.

```
fakeAPI := &api{
  id:         "fakeAPI",
}

fakeAPI.getSetStateOptions(...)
```